File: PO/Java-C-performance-comparison/src/c/shortest_path.c
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>

#define INF 1000000009

struct __Graph;
typedef struct __Graph Graph;

Graph* createGraph(int);
void addEdge(Graph*, int, int, int);
int getEdge(Graph*, int, int);

Graph* readGraph();
int **algorithmFloydWarshall(Graph*);
void answerQuedy(int**);

int main(void) {
    Graph *graph = readGraph();
    int **shortest = algorithmFloydWarshall(graph);
    answerQuedy(shortest);

    return 0;
}

void *safeMalloc(size_t size) {
    void *ptr = malloc(size);

    if (ptr == NULL) {
        fprintf(stderr, "Memory outage.\n");
        exit(1);
    }

    return ptr;
}

int **createTwoDimArray(int X, int Y) {
    int **array = safeMalloc(sizeof(int*) * X);
    for (int i = 0; i < X; i++) {
        array[i] = safeMalloc(sizeof(int) * Y);
    }

    return array;
}

int min(int a, int b) {
    return a < b ? a : b;
}

Graph* readGraph() {
    int N, M, A, B, L;
    if (scanf(" %d %d", &N, &M) != 2)
        exit(1);

    Graph *graph = createGraph(N);

    while (M --> 0) {
        if (scanf(" %d %d %d", &A, &B, &L) != 3)
            exit(1);

        assert(0 <= A && A < N);
        assert(0 <= B && B < N);
        assert(0 <= L && L < INF);

        addEdge(graph, A, B, L);
    }

    return graph;
}

struct __Graph {
    int N; ///< Number of nodes.
    int **dist; ///< Distance matrix.
};

Graph* createGraph(int N) {
    Graph* graph = safeMalloc(sizeof(Graph));

    graph->N = N;
    graph->dist = createTwoDimArray(N, N);
    for (int i = 0; i < N; i++) {

        for (int j = 0; j < N; j++) {
            graph->dist[i][j] = INF;
        }
        graph->dist[i][i] = INF;
    }

    return graph;
}

void addEdge(Graph *graph, int begin, int end, int len) {
    graph->dist[begin][end] = min(graph->dist[begin][end], len);
}

int getEdge(Graph *graph, int begin, int end) {
    return graph->dist[begin][end];
}

int **algorithmFloydWarshall(Graph *graph) {
    int N = graph->N;
    int **shortest = createTwoDimArray(N, N);

    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++) {
            shortest[i][j] = getEdge(graph, i, j);
        }
    }

    for (int k = 0; k < N; k++) {
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (shortest[i][j] > shortest[i][k] + shortest[k][j]) {
                    shortest[i][j] = shortest[i][k] + shortest[k][j];
                }
            }
        }
    }

    return shortest;
}

void answerQuedy(int **shortest) {
    int queries, A, B, answer;
    if (scanf(" %d", &queries) != 1)
        exit(1);

    while (queries --> 0) {
        if (scanf(" %d %d", &A, &B) != 2)
            exit(1);
            
        answer = shortest[A][B];

        if (answer == INF)
            printf("INF\n");
        else
            printf("%d\n", answer);
    }
}

File: PO/Java-C-performance-comparison/src/ocaml/shortest_path.ml

(* Graph with size and matrix of distances. *)
type graph = { size : int; distances : int array array }


let inf = 1000000009


let makeGraph n = 
  let dist = Array.make_matrix n n inf
  in 
  for i = 0 to n - 1 do
    Array.set (Array.get dist i) i 0
  done;
  { size = n;  distances = dist }


let getEdge { size = size; distances = dist } beg en = 
  Array.get (Array.get dist beg) en


let addEdge ({ size = size; distances = dist } as g) beg en len =
  let actualLen = getEdge g beg en in
  if actualLen > len then
    Array.set (Array.get dist beg) en len


let getN { size = size } = size


let readGraph () = 
  let n = ref 0 in
  let m = ref 0 in
  Scanf.scanf " %d %d" (fun x y -> n := x; m := y);
  let n = !n in
  let m = !m in
  let g = makeGraph n in

  for i = 0 to m - 1 do
    let a = ref 0 in
    let b = ref 0 in
    let l = ref 0 in
    Scanf.scanf " %d %d %d" (fun x y z -> a := x; b := y; l := z);
    let a = !a in
    let b = !b in
    let l = !l in

    assert (0 <= a && a < n);
    assert (0 <= b && b < n);
    assert (0 <= l && l < inf);

    addEdge g a b l
  done; g


let algorithmFloydWarshall g = 
  let n = getN g in
  let shortest = Array.make_matrix n n 0 in
  
  for i = 0 to n - 1 do
    for j = 0 to n - 1 do
      Array.set (Array.get shortest i) j (getEdge g i j)
    done
  done;

  let get arr x y = Array.get (Array.get arr x) y in

  for k = 0 to n - 1 do
    for i = 0 to n - 1 do
      for j = 0 to n - 1 do
        let newPath = get shortest i k + get shortest k j in
        if get shortest i j > newPath then
          Array.set (Array.get shortest i) j newPath
      done
    done
  done;

  shortest


let answerQuery shortest = 
  let q = ref 0 in
  Scanf.scanf " %d" (fun x -> q := x);
  let q = !q in 
  for i = 0 to q - 1 do
    let a = ref 0 in
    let b = ref 0 in
    Scanf.scanf " %d %d" (fun x y -> a := x; b := y);
    let a = !a in
    let b = !b in
    let anwser = Array.get (Array.get shortest a) b in
    if anwser == inf then
      Printf.printf "INF\n"
    else
      Printf.printf "%d\n" anwser
  done


let main () = 
  let g = readGraph () in
  let shortest = algorithmFloydWarshall g in
  answerQuery shortest
;;

main ()
;;